signature/mechanism_sigstore.go

```diff
@@ -0,0 +1,325 @@
+	"context"
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/sha256"
+	"crypto/x509"
+	"encoding/hex"
+	"encoding/json"
+	"encoding/pem"
+	"fmt"
+	"net/url"
+	"os"
+	"strings"
+
+	"github.com/cyberphone/json-canonicalization/go/src/webpki.org/jsoncanonicalizer"
+	httptransport "github.com/go-openapi/runtime/client"
+	"github.com/go-openapi/strfmt"
+	"github.com/go-openapi/swag"
+	"github.com/google/trillian/merkle/logverifier"
+	"github.com/google/trillian/merkle/rfc6962/hasher"
+	"github.com/pkg/errors"
+	fulcioClient "github.com/sigstore/fulcio/pkg/client"
+	"github.com/sigstore/fulcio/pkg/generated/client/operations"
+	fulcioModels "github.com/sigstore/fulcio/pkg/generated/models"
+	rekorClient "github.com/sigstore/rekor/pkg/client"
+	"github.com/sigstore/rekor/pkg/generated/client"
+	"github.com/sigstore/rekor/pkg/generated/client/entries"
+	"github.com/sigstore/rekor/pkg/generated/client/pubkey"
+	"github.com/sigstore/rekor/pkg/generated/models"
+	rekord_v001 "github.com/sigstore/rekor/pkg/types/rekord/v0.0.1"
+	"github.com/sigstore/sigstore/pkg/cryptoutils"
+	"github.com/sigstore/sigstore/pkg/oauthflow"
+	"github.com/sigstore/sigstore/pkg/signature"
+	"github.com/sigstore/sigstore/pkg/signature/options"
+)
+
+const (
+	defaultRekorURL  = "https://rekor.sigstore.dev"
+	defaultFulcioURL = "https://fulcio.sigstore.dev"
+	oidcIssuer       = "https://oauth2.sigstore.dev/auth"
+	oidcClientID     = "sigstore"
+)
+
+func rekorServer() string {
+	// TODO: Use another means e.g. flag, policy, etc.
+	addr := os.Getenv("REKOR_URL")
+	if addr != "" {
+		return addr
+	}
+	return defaultRekorURL
+}
+
+func fulcioServer() string {
+	// TODO: Use another means e.g. flag, policy, etc.
+	addr := os.Getenv("FULCIO_URL")
+	if addr != "" {
+		return addr
+	}
+	return defaultFulcioURL
+}
+
+type sigstoreSigningMechanism struct {
+	ctx   context.Context
+	cert  []byte
+	chain []byte
+	*signature.ECDSASignerVerifier
+}
+
+// newSigstoreSigningMechanism returns a new sigstore signing mechanism.
+// The caller must call .Close() on the returned SigningMechanism.
+func newSigstoreSigningMechanism() (SigstoreSigningMechanism, error) {
+	return &sigstoreSigningMechanism{
+		ctx: context.Background(),
+	}, nil
+}
+
+// Close removes resources associated with the mechanism, if any.
+func (s *sigstoreSigningMechanism) Close() error {
+	return nil
+}
+
+// SupportsSigning returns nil if the mechanism supports signing, or a SigningNotSupportedError.
+func (s *sigstoreSigningMechanism) SupportsSigning() error {
+	return nil
+
+}
+
+func (s *sigstoreSigningMechanism) GenerateCertificate() error {
+	fulcioServer, err := url.Parse(fulcioServer())
+	if err != nil {
+		return errors.Wrap(err, "parsing Fulcio URL")
+	}
+
+	fClient := fulcioClient.New(fulcioServer)
+	if fClient == nil {
+		return errors.New("error initializing fulcio client")
+	}
+
+	// TODO: use debug log
+	fmt.Println("Generating ephemeral keys")
+	priv, err := generatePrivateKey()
```

Comment:
Besides being a good practice, AFAIK FIPS requires explicit zeroization of all secret material after it is no longer used. How is that going to work?

`crypto/ecdsa` doesn’t seem to attempt to do this at all.

(Yeah, it’s hard. That’s why the previous implementation only allowed signing via a separate third-party non-Go process.)